src/Microsoft.SymbolStore/ChecksumValidator.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.FileFormats.PE;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.SymbolStore
+{
+    class ChecksumValidator
+    {
+        private const string pdbStreamName = "#Pdb";
+        private const uint pdbIdSize = 20;
+
+        internal static void Validate(ITracer tracer, Stream pdbStream, IEnumerable<PdbChecksum> pdbChecksums)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                pdbStream.CopyTo(ms);
+                byte[] bytes = ms.ToArray();
+                uint offset = 0;
+
+                try
+                {
+                    offset = GetPdbStreamOffset(pdbStream);
+                }
+                catch (Exception ex)
+                {
+                    tracer.Error(ex.Message);
+                    throw;
+                }
+
+                for (int i = 0; i <= pdbIdSize; i++)
+                {
+                    bytes[i + offset] = 0;
+                }
+
+                foreach (var checksum in pdbChecksums)
+                {
+                    tracer.Information($"Testing checksum: {checksum}");
+
+                    var algorithm = HashAlgorithm.Create(checksum.AlgorithmName);
+                    if (algorithm != null)
+                    {
+                        var hash = algorithm.ComputeHash(bytes);
+                        if (hash.SequenceEqual(checksum.Checksum))
+                        {
+                            // If any of the checksums are OK, we're good
+                            tracer.Information($"Found checksum match {checksum}");
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Unknown hash algorithm");
+                    }
+                }
+            }
+            throw new InvalidChecksumException("PDB checksum mismatch");
+        }
+
+        private static uint GetPdbStreamOffset(Stream pdbStream)
+        {
+            pdbStream.Position = 0;
+            var reader = new BinaryReader(pdbStream);
```

Comment:
I believe this will close the pdbStream when the BinaryReader is finialized. Do you want the overload with the `leaveOpen` argument?

Reply:
TIL